### PR TITLE
Refactor async executor for more trainer and worker configuration, add fleet_wrapper for pslib interface

### PR DESCRIPTION
**Pull request by guru4elephant**

### CODE REVIEW

File: paddle/fluid/framework/trainer_factory.cc [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+std::shared_ptr<TrainerBase> TrainerFactory::CreateTrainer(
+    std::string trainer_class) {
+  if (g_trainer_map.count(trainer_class) < 1) {
+    exit(-1);
+  }
+  return g_trainer_map[trainer_class]();
+}
+
+REGISTER_TRAINER_CLASS(MultiTrainer);
+REGISTER_TRAINER_CLASS(DistMultiTrainer);
```
Comment (wangguibao) [inline paddle/fluid/framework/trainer_factory.cc]: Maybe better to put REGISTER_TRAINER_CLASS(MultiTrainer) to the end of each MultiTrainer.cc, REGISTER_TRAINER_CLASS(DistMultiTrainer) to the end of DistMultiTrainer.cc

Comment (wangguibao) [inline paddle/fluid/framework/trainer_factory.cc]: Maybe better to put REGISTER_TRAINER_CLASS(MultiTrainer) to the end of multi_trainer.cc, REGISTER_TRAINER_CLASS(DistMultiTrainer) to the end of dist_multi_trainer.cc

File: paddle/fluid/framework/async_executor.h [hunk 2]
```diff
@@ -87,21 +85,14 @@ class AsyncExecutor {
                      const std::vector<std::string>& fetch_var_names,
                      Scope* root_scope, const int thread_index,
                      const bool debug);
```
Comment (wangguibao) [inline paddle/fluid/framework/async_executor.h]: It seems `AsyncExecutor::CreateThreads()` is not used anymore

File: paddle/fluid/framework/device_worker.h [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+  PullDenseWorker() {}
+  virtual ~PullDenseWorker() {}
+  virtual void Initialize(const TrainerDesc& param);
+  int Start();
+  void Stop();
+  void SetScope(Scope* scope) { root_scope_ = scope; }
+  void IncreaseThreadVersion(int thread_id, uint64_t table_id);
+  void ResetThreadVersion(uint64_t table_id);
+  void Wait(std::vector<::std::future<int32_t>>* status_vec);
+  static std::shared_ptr<PullDenseWorker> s_instance_;
```
Comment (wangguibao) [inline paddle/fluid/framework/device_worker.h]: `static std::shared_ptr<PullDenseWorker> s_instance_;`
Shall we make `s_instance_` a static local variable inside `GetInstance()`, or make it a private member of `PullDenseWorker`, to ensure a singleton

File: paddle/fluid/framework/device_worker.h [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+#include "paddle/fluid/operators/reader/blocking_queue.h"
+#include "paddle/fluid/platform/place.h"
+#include "paddle/fluid/platform/timer.h"
+
+namespace paddle {
+namespace framework {
+
+class PullDenseWorker {
+ public:
+  PullDenseWorker() {}
```
Comment (wangguibao) [inline paddle/fluid/framework/device_worker.h]: Shall we make this default constructor a private member of class `PullDenseWorker`, to ensure a singleton

File: paddle/fluid/framework/device_worker.h [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+  // feasign stats
+  std::map<uint64_t, std::vector<float>> feature_labels_;
+  // feasign embedding
+  std::map<uint64_t, std::vector<std::vector<float>>> feature_values_;
+  // feasign embedding gradient
+  std::map<uint64_t, std::vector<std::vector<float>>> feature_grads_;
+  // skipped ops
+  std::vector<std::string> skip_ops_;
+
+  std::shared_ptr<PullDenseWorker> _pull_dense_worker;
```
Comment (wangguibao) [inline paddle/fluid/framework/device_worker.h:182]: Is this member `_pull_dense_worker` typo error, or duplicate with that of Line 158?

File: paddle/fluid/framework/device_worker_factory.cc [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+std::shared_ptr<DeviceWorker> DeviceWorkerFactory::CreateDeviceWorker(
+    std::string device_worker_class) {
+  if (g_device_worker_map.count(device_worker_class) < 1) {
+    exit(-1);
+  }
+  return g_device_worker_map[device_worker_class]();
+}
+
+REGISTER_DEVICE_WORKER_CLASS(HogwildWorker);
+REGISTER_DEVICE_WORKER_CLASS(DownpourWorker);
```
Comment (wangguibao) [inline paddle/fluid/framework/device_worker_factory.cc]: Maybe better to put `REGISTER_DEVICE_WORKER_CLASS(HogwildWorker)` at end of hogwild_worker.cc, put `REGISTER_DEVICE_WORKER_CLASS(DownpourWorker)` to the end of downpour_worker.cc

File: paddle/fluid/framework/fleet/fleet_wrapper.h [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+  void InitServer(const std::string& dist_desc, int index);
+  void InitWorker(const std::string& dist_desc,
+                  const std::vector<uint64_t>& host_sign_list, int node_num,
+                  int index);
+  void StopServer();
+  uint64_t RunServer();
+  void GatherServers(const std::vector<uint64_t>& host_sign_list, int node_num);
+
+  static std::shared_ptr<FleetWrapper> s_instance_;
+  static std::shared_ptr<FleetWrapper> GetInstance() {
```
Comment (wangguibao) [inline paddle/fluid/framework/fleet/fleet_wrapper.h:115]: Maybe we should put `static std::shared_ptr<FleetWrapper> s_instance_` a static local variable inside `GetInstance()`, to ensure a singleton

File: paddle/fluid/framework/fleet/fleet_wrapper.h [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+// Push
+//   Sync: PushSparseVarsSync
+//   Async: PushSparseVarsAsync
+// Push dense variables to server in Async mode
+// Param<in>: scope, table_id, var_names
+// Param<out>: push_sparse_status
+
+class FleetWrapper {
+ public:
+  FleetWrapper() {}
```
Comment (wangguibao) [inline paddle/fluid/framework/fleet/fleet_wrapper.h:51]: Maybe we should make this default constructor a private member of `FleetWrapper`, to ensure a singleton

File: paddle/fluid/framework/pull_dense_worker.cc [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+  return 0;
+}
+
+void PullDenseWorker::Run() {
+  while (running_) {
+    pull_dense_status_.resize(0);
+    for (size_t i = 0; i < param_.dense_table_size(); ++i) {
+      uint64_t tid = static_cast<uint64_t>(param_.dense_table(i).table_id());
+      if (CheckUpdateParam(tid)) {
+        fleet_ptr_->PullDenseVarsAsync(
```
Comment (wangguibao) [inline paddle/fluid/framework/pull_dense_worker.cc:82]: `fleet_ptr_` seems not initialized before used

File: paddle/fluid/framework/trainer_factory.cc (hunk 1 above)
Comment (guru4elephant) [reply paddle/fluid/framework/trainer_factory.cc]: done.

File: paddle/fluid/framework/device_worker.h (hunk 4 above)
Comment (guru4elephant) [reply paddle/fluid/framework/device_worker.h]: done.

File: paddle/fluid/framework/device_worker.h (hunk 5 above)
Comment (guru4elephant) [reply paddle/fluid/framework/device_worker.h:182]: fixed

File: paddle/fluid/framework/device_worker_factory.cc (hunk 6 above)
Comment (guru4elephant) [reply paddle/fluid/framework/device_worker_factory.cc]: done.

File: paddle/fluid/framework/fleet/fleet_wrapper.h (hunk 7 above)
Comment (guru4elephant) [reply paddle/fluid/framework/fleet/fleet_wrapper.h:115]: done.

File: paddle/fluid/framework/fleet/fleet_wrapper.h (hunk 8 above)
Comment (guru4elephant) [reply paddle/fluid/framework/fleet/fleet_wrapper.h:51]: done.

File: paddle/fluid/framework/pull_dense_worker.cc (hunk 9 above)
Comment (guru4elephant) [reply paddle/fluid/framework/pull_dense_worker.cc:82]: fixed

File: paddle/fluid/framework/async_executor.h (hunk 2 above)
Comment (guru4elephant) [reply paddle/fluid/framework/async_executor.h]: removed

File: paddle/fluid/framework/device_worker.h (hunk 3 above)
Comment (guru4elephant) [reply paddle/fluid/framework/device_worker.h]: fixed

**Pull request closed without merging**